src/emlid_reach.cpp

```diff
@@ -0,0 +1,391 @@
+#include <string.h>
+#include <ctime>
+#include <cmath>
+#include <cstring>
+#include <errno.h>
+
+#include <stdlib.h>
+
+//// ERB
+// 'E' Ox45 | 'R' Ox52 | ID | LENGTH (2 Bytes little endian) | payload | CHECKSUM (2B)
+#define ERB_SYNC_1 0x45 // E
+#define ERB_SYNC_2 0x52 // R
+
+#define ERB_ID_VERSION         0x01
+#define ERB_ID_GEODIC_POSITION 0x02
+#define ERB_ID_NAV_STATUS      0x03
+#define ERB_ID_DOPS            0x04
+#define ERB_ID_VELOCITY_NED    0x05
+#define ERB_ID_SPACE_INFO      0x06 // not used, reduces stack usage
+#define ERB_ID_RTK             0x07 // RTK, TODO
+
+
+#define EMLID_UNUSED(x) (void)x;
+
+#define GPS_PI 3.141592653589793238462643383280f
+
+#define AUTO_DETECT_MAX_TIMEOUT       50 // if more detect failed AND
+#define AUTO_DETECT_MAX_PARSE_ERR     2  // if more detect failed AND
+#define AUTO_DETECT_MAX_READ_SENTENCE 10 // if more detect succeed
+
+
+GPSDriverEmlidReach::GPSDriverEmlidReach(GPSCallbackPtr callback, void *callback_user,
+		struct vehicle_gps_position_s *gps_position, struct satellite_info_s *satellite_info) :
+	GPSHelper(callback, callback_user),
+	_gps_position(gps_position), _satellite_info(satellite_info)
+{}
+
+
+int
+GPSDriverEmlidReach::configure(unsigned &baudrate, OutputMode output_mode)
+{
+	// TODO RTK
+	if (output_mode != OutputMode::GPS) {
+		GPS_WARN("EMLIDREACH: Unsupported Output Mode %i", (int)output_mode);
+		return -1;
+	}
+
+	unsigned baud_allowed[] {57600, 115200, 230400};
+
+	for (unsigned i = 0; i < sizeof(baud_allowed) / sizeof(baud_allowed[0]); i++) {
+		if (baudrate > 0 && baudrate != baud_allowed[i]) {
+			continue;
+		}
+
+		_sentence_cnt = 0;
+		_parse_err_cnt = 0;
+		_erb_decode_state  = ERB_State::init;
+		_erb_payload_len = 0;
+		_erb_buff_cnt = 0;
+
+		if (GPSHelper::setBaudrate(baud_allowed[i]) != 0) {
+			continue;
+		}
+
+		if (! testConnection()) {
+			continue;
+		}
+
+		GPS_INFO("EmlidReach: config OK with ERB baudrate: %d", baud_allowed[i]);
+		baudrate = baud_allowed[i];
+		return 0;
+	}
+
+	return -1;
+}
+
+
+bool
+GPSDriverEmlidReach::testConnection()
+{
+	_testing_connection = true;
+
+	unsigned timeout_cnt = 0;
+
+	while (timeout_cnt < AUTO_DETECT_MAX_TIMEOUT
+	       && _parse_err_cnt < AUTO_DETECT_MAX_PARSE_ERR
+	       && _sentence_cnt < AUTO_DETECT_MAX_READ_SENTENCE) {
+		if (receive(100) == 0b100) {
+			timeout_cnt ++;
+		}
+	}
+
+	_testing_connection = false;
+	return _sentence_cnt >= AUTO_DETECT_MAX_READ_SENTENCE;
+}
+
+
+int
+GPSDriverEmlidReach::receive(unsigned timeout)
+{
+	uint8_t read_buff[GPS_READ_BUFFER_SIZE];
+	unsigned read_buff_len = 0;
+	uint8_t read_ind = 0;
+	// 1 if gps_position updated, 2 for satellite_info_s (can be bit OR), 0b100 for nothing
+	int return_status = 0b100;
+
+	// read from serial
+	read_buff_len = read(read_buff, sizeof(read_buff), timeout);
+
+	if (read_buff_len > 0) {
+		read_ind = 0;
+
+	} else {
+		// timeout occured
+		return 0b100; // no update to publish, but for testConnection() to know about timeout
```

Comment:
This is not conforming with the API definition of `receive()`. We either need to update the API spec and all users of it (QGC and the Firmware), or rather change this to be conforming.

Reply:
@bkueng Yes, indeed. The problem I tried to solve here is that it cannot return 0, or else the loop in gps.cpp exits (line 750)
`while ((helper_ret = _helper->receive(TIMEOUT_5HZ)) > 0 && !should_exit()) {`
I do not have access to should_exit() either
And I didn't want to loop until I have something to receive, or else that could block the gps thread into the driver
What ubx is doing is to exist after 1000 timeout (ubx.cpp line 762)
`/* abort after timeout if no useful packets received */
if (time_started + timeout * 1000 < gps_absolute_time()) {
	UBX_DEBUG("timed out, returning");
	return -1;
}`
I thought that returning a non zero value without the bit set signalling gps or sat update would be more elegant
I also considered changing gps.cpp to
`while ((helper_ret = _helper->receive(TIMEOUT_5HZ)) >= 0 && !should_exit()) {`
But I didn't want to change things outside emlid files

Anyways, I'm changing that to how it is in ubx/ashtech/mtk now